tasks/eva_2205/src/main/groovy/eva2205/eva2205_undo_merges_into_multimaps.groovy

```diff
@@ -0,0 +1,213 @@
+     * See https://docs.google.com/spreadsheets/d/1kRKHR4zrq-nxH_Sg82TwXZbxdgjB3K-q4YMyJva8yMg/edit#rangeid=1814908308
+     * @param recentSvoeOpsWithoutMapWt - Most recent SVOE/dbsnpSVOE records without map weight for a set of SS hashes (keyed by SS hash)
+     * @param currentSSRecords - Current SS record for a set of SS hashes (keyed by SS hash)
+     * @return None
+     */
+    void undoMultiMapRSAssignment(Map<String, List> recentSvoeOpsWithoutMapWt, Map<String, List> currentSSRecords) {
+        def (bulkSVEUpdateOps, bulkdbSnpSVEUpdateOps) =
+        [sveClass, dbsnpSveClass].collect { dbEnv.mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED, it) }
+        def (svoeInserts, dbSnpSvoeInserts, dbsnpCvoeInserts) = [new ArrayList<>(), new ArrayList<>(),
+                                                                 new ArrayList<>()]
+        def (numSVEUpdates, numDbsnpSveUpdates) = [0, 0]
+
+        recentSvoeOpsWithoutMapWt.each { ssHash, recentSSOpWithoutMapWt ->
+            def ssRecordToUpdate = currentSSRecords.get(ssHash)[0]
+            def oldRS = ssRecordToUpdate.clusteredVariantAccession
+            // Since we are undoing a previous RS assignment to this SS,
+            // the new RS is the one from the Op recording that previous assignment
+            def newRS = recentSSOpWithoutMapWt.inactiveObjects[0].clusteredVariantAccession
+            Query queryToFindSSToUpdate = query(where("_id").is(ssHash).and("rs").ne(newRS))
+
+            def ssAccession = ssRecordToUpdate.accession
+            def isEvaSS = (ssAccession >= dbsnpSSIDUpperBound)
+            def (sveUpdateObj, svoeInsertObj) = isEvaSS ?
+                    [bulkSVEUpdateOps, svoeInserts] : [bulkdbSnpSVEUpdateOps, dbSnpSvoeInserts]
+
+            logger.info("SS with hash ${ssHash} and accession ${ssAccession} will be assigned " +
+                    "non-map weighted RS ${newRS}...")
+            def updatesToMake = new Update()
+            updatesToMake.set("rs", newRS)
+            updatesToMake.unset("mapWeight")
+            sveUpdateObj.updateOne(queryToFindSSToUpdate, updatesToMake)
+
+            def svoeOpToRecordUnmerge = new SubmittedVariantOperationEntity()
+            svoeOpToRecordUnmerge.fill(EventType.UPDATED, ssAccession, "Undo map-weighted RS assignment.",
+                    Arrays.asList(new SubmittedVariantInactiveEntity(ssRecordToUpdate)))
+            svoeOpToRecordUnmerge.setId("UNDO_MULTIMAP_RS_ASSIGN_${ssRecordToUpdate.hashedMessage}")
+            svoeInsertObj.add(svoeOpToRecordUnmerge)
+            if (isEvaSS) {
+                numSVEUpdates += 1
+            } else {
+                numDbsnpSveUpdates += 1
+            }
+
+            def cvoeOpToRecordUnmerge = new ClusteredVariantOperationEntity()
+            def oldCVE = EVAObjectModelUtils.toClusteredVariantEntity(ssRecordToUpdate)
+            cvoeOpToRecordUnmerge.fill(EventType.MERGED, oldRS, newRS, "Undo merge into map-weighted RS.",
```

Comment:
Are we using the MERGE event type to record the UNDO merge ?
Isn't that going to be confusing ?

Reply:
It is a merge "in spirit" since we are doing similar actions as we would do in a merge but with the source and target being reversed. I wasn't too sure about creating yet another event definition. We can discuss this further during the next standup...

Reply:
After discussion during standup, it looks like these merge records could be problematic and could be misleadingly accounted for in merged RS in the release files